src/object.d

```diff
@@ -3722,6 +3722,186 @@ template _arrayOp(Args...)
+            if (caseLabels[i].length == condition.length)
+            {
+                cj = __cmp(caseLabels[i], condition);
+                if (cj == 0)
+                {
+                    assert(i == result);
+                    break;
+                }
+            }
+        }
+    }
+}
+body
+{
+    size_t low  = 0;
+    size_t high = caseLabels.length;
+
+    if (high &&
+        condition.length >= caseLabels[0].length &&
+        condition.length <= caseLabels[high - 1].length)
+    {
+        // Looking for 0 length string, which would only be at the beginning
+        if (condition.length == 0)
+            return 0;
+
+        T c1 = condition[0];
+
+        // Do binary search
+        while (low < high)
+        {
+            auto mid = (low + high) >> 1;
+            auto pca = caseLabels[mid];
+            auto c   = cast(sizediff_t)(condition.length - pca.length);
+            if (c == 0)
+            {
+                c = cast(ubyte)c1 - cast(ubyte)pca[0];
+                if (c == 0)
+                {
+                    c = __cmp(condition, pca);
+                    if (c == 0)
+                    {
+                        return cast(int)mid;
+                    }
+                }
+            }
+            if (c < 0)
+            {
+                high = mid;
+            }
+            else
+            {
+                low = mid + 1;
+            }
+        }
+    }
+
+    return -1; // not found
+}
+
+unittest
+{
+    switch (cast(char []) "c")
+    {
+         case "coo":
+         default:
+             break;
+    }
+
+    int bug5381(string s)
+    {
+        switch(s)
+        {
+            case "unittest":        return 1;
+            case "D_Version2":      return 2;
+            case "none":            return 3;
+            case "all":             return 4;
+            default:                return 5;
+        }
+    }
+    int rc = bug5381("none");
+    assert(rc == 3);
+}
+
+unittest
+{
+    switch (cast(wchar []) "c")
+    {
+         case "coo":
+         default:
+             break;
+    }
+
+    int bug5381(wstring ws)
+    {
+        switch(ws)
+        {
+            case "unittest":        return 1;
+            case "D_Version2":      return 2;
+            case "none":            return 3;
+            case "all":             return 4;
+            default:                return 5;
+        }
+    }
+    int rc = bug5381("none"w);
+    assert(rc == 3);
+}
+
+unittest
+{
+    switch (cast(dchar []) "c")
+    {
+         case "coo":
+         default:
+             break;
+    }
+
+    int bug5381(dstring ds)
+    {
+        switch(ds)
+        {
+            case "unittest":        return 1;
+            case "D_Version2":      return 2;
+            case "none":            return 3;
+            case "all":             return 4;
+            default:                return 5;
+        }
+    }
+    int rc = bug5381("none"d);
+    assert(rc == 3);
+}
+
```

Comment:
Looks like more unittests are needed given that the bugs involving lengths were not caught.